File: Algorithms/087.ScrambleString.py
class Solution(object):
    def isScramble(self, s1, s2):
        """
        :type s1: str
        :type s2: str
        :rtype: bool
        """
        l1, l2 = len(s1), len(s2)
        if l1 != l2:
            return False
        if l1 == 0:
            return True
        if l1 == 1:
            return s1 == s2
        ss1, ss2 = sorted(s1), sorted(s2)
        if ss1 != ss2:
            return False
        for i in range(1, l1):
            s11, s12 = s1[:i], s1[i:]
            s21, s22 = s2[:i], s2[i:]
            if self.isScramble(s11, s21) and self.isScramble(s12, s22):
                return True
            s23, s24 = s2[:l1-i], s2[l1-i:]
            if self.isScramble(s11, s24) and self.isScramble(s12, s23):
                return True
        return False

File: Algorithms/087.ScrambleString.cpp
class Solution {
public:
    bool isScramble(string s1, string s2) {
        int l1 = s1.length();
        int l2 = s2.length();
        if (l1 != l2)
            return false;
        if (l1 == 0)
            return true;
        if (l1 == 1) 
            return s1 == s2;
        string s11 = s1;
        string s22 = s2;
        sort(s11.begin(), s11.end());
        sort(s22.begin(), s22.end());
        if (s11 != s22)
            return false;
        for (int i=1; i < l1; i++) {
            string s11 = s1.substr(0, i);
            string s12 = s1.substr(i);
            string s21 = s2.substr(0, i);
            string s22 = s2.substr(i);
            if (isScramble(s11, s21) && isScramble(s12, s22))
                return true;
            string s23 = s2.substr(0, l1-i);
            string s24 = s2.substr(l1-i);
            if (isScramble(s11, s24) && isScramble(s12, s23))
                return true;
        }
        return false;
    }
};
